src/irc/Ident.ts

```diff
@@ -130,19 +130,26 @@ class IdentSrv {
             this.respond(sock, localPort, remotePort, username);
             return;
         }
+        const mappingHash = Object.keys(this.portMappings).join("|");
         // Wait for pending connections to finish first.
         await Promise.all([...this.pendingConnections]);
+        // We don't know here if anything has actually changed, so compare hashes.
+        if (Object.keys(this.portMappings).join("|") !== mappingHash) {
+            // Hash has changed, retry.
+            await this.tryRespond(sock, localPort, remotePort);
```

Comment:
I don't think recursion is the way to go here. This code can recurse forever now as long as new pending connections are made.

The old code avoided recursing forever by carrying the `currentToken` through the recursion. This way it only recurses until the pending connections *at the start of recursion* are completed.

Your `await` does all that in one step, though. So I would write it like this:

```typescript
    private async respond(sock: net.Socket, localPort: string, remotePort: string) {
        let username = this.portMappings[localPort];
        if (!username) {
            // Wait for pending connections to finish first.
            await Promise.all([...this.pendingConnections]);
            username = this.portMappings[localPort];
        }

        let response;
        if (username) {
            log.debug("Port %s is %s", localPort, username);
            response = `${localPort},${remotePort}:USERID:UNIX:${username}\r\n`;
        } else {
            log.debug("No user on port %s", localPort);
            response = `${localPort},${remotePort}:ERROR:NO-USER\r\n`;
        }
        log.debug(response);
        sock.end(response);
    }
```

This also subsumes `respond` into `tryRespond` and renames the latter to the former.

Reply:
Yeah, I can't fault that. Will try it.